### DISCUSSION TITLE
New Piece: Traveling Vagabond

### DISCUSSION THREAD

**Discussion by Xavman42** [DigiScore/neoscore#106] [Show and tell]
Just finished a video-score using neoscore! The score was rendered as a video file then uploaded [here](https://youtu.be/oyae_aYv4hI). The rather messy repository is [here](https://github.com/Xavman42/traveling_vagabond). Some notes on the piece can be found in the YouTube video description.
The piece will be performed in early December and I'll hopefully have a video recording shortly after that!

**Comment by craigvear** [comment]
Wow. An amazing feat. It runs very cleanly on YT. How could Neoscore support this better?

**Reply by Xavman42** [reply]
I think my dream feature, one that would really expedite making scrolling scores like this, would be the ability to superimpose multiple documents in the same viewport. Most of my animated ideas so far would benefit from cel-style animation like in the early Disney films. If I could choose a document for an object to be rendered on, then I could overlay an arbitrary number of transparent documents and shift them around relative to each other. This would make designing a HUD like in this piece and in GuitaRPG trivial. And if re-renders happen at the document level (I'm pretty sure that's how that works?), maybe different documents could be rendered independently, meaning something could be animated on one document layer without needing to re-render everything else on the screen.

And of course, I need to finish implementing native video rendering at some point...

**Comment by ajyoon** [comment]
Very cool! I checked out your code as well and tried to see if I could make your life easier, but it seems to be more than a 10 minute job.. writing temp images to disk before sending to QT is definitely not great! Neoscore could pretty easily be updated to emit raw bitmaps in-memory for immediate sending to other processes (cv2 or, as I think will probably be better, ffmpeg). I also noticed the overlay jiggles around a bit, I guess caused by positioning rounding errors, and the resulting true frame rate for some reason is not 60 as the code specifies.

**Reply by craigvear** [reply]
You are the oracle Andrew.

**Reply by Xavman42** [reply]
Thanks! I'm really not doing myself any favors with my coding style - I'm trying to make my code more readable with every project.

The way I'm rendering right now is inefficient to say the least. With a target output resolution of 4K at 60 fps, it takes about an hour to render a minute of this score on my laptop. I'm sure ffmpeg will be better too, I was being lazy here and just rendering overnight...

I also suspect rounding errors are causing the jiggling. I wonder if that issue may be exacerbated by my working mostly in Units (as opposed to Mm or Inches).

**Reply by ajyoon** [reply]
The viewport jiggling might actually be caused by neoscore itself. `neoscore.get_viewport_center_pos` is a bit of a black box since it works by asking Qt to work out the viewport position by reversing its transform matrix. There's plenty of room for rounding errors to compound along the line there. A quick glance at your code doesn't seem to reveal anything that would be a problem. Under the hood neoscore units do all arithmetic on underlying float equivalents with full precision; rounding only comes into play with comparisons and string serialization.

**Reply by ajyoon** [reply]
(which is to say, it shouldn't matter what Unit type you use, all should be equally precise except when printed)

**Comment by Xavman42** [comment]
The piece has premiered!

https://youtu.be/G7dVgWk7-Qo

There was a bit of a synchronization issue with the premiere, so the piano is nearly a second behind the strings for the duration of the piece, but it still works pretty well regardless! I may get access to a dress-rehearsal recording later which is better synchronized. The piece was extremely well received by an audience of ~100 people here at Illinois.

**Reply by ajyoon** [reply]
Awesome stuff, congratulations! It sounds great, even without considering all the interesting compositional and notational techniques that went into it. Very well done!
